src/ucp/core/ucp_worker.c

```diff
@@ -1730,29 +1727,35 @@ static ucs_status_t ucp_worker_init_mpools(ucp_worker_h worker)
         }
     }
 
-    /* Create memory pool for incoming UCT messages without a UCT descriptor */
-    status = ucs_mpool_init(&worker->am_mp, 0,
-                            max_mp_entry_size + UCP_WORKER_HEADROOM_SIZE +
-                            worker->am.alignment,
-                            0, UCS_SYS_CACHE_LINE_SIZE, 128, UINT_MAX,
-                            &ucp_am_mpool_ops, "ucp_am_bufs");
-    if (status != UCS_OK) {
-        goto err_rkey_mp_cleanup;
-    }
-
     /* Create memory pool of bounce buffers */
     status = ucs_mpool_init(&worker->reg_mp, 0,
                             context->config.ext.seg_size + sizeof(ucp_mem_desc_t),
                             sizeof(ucp_mem_desc_t), UCS_SYS_CACHE_LINE_SIZE,
                             128, UINT_MAX, &ucp_reg_mpool_ops, "ucp_reg_bufs");
     if (status != UCS_OK) {
-        goto err_am_mp_cleanup;
+        goto err_rkey_mp_cleanup;
     }
 
+    if (max_mp_entry_size > 0) {
+        /* Create memory pool for incoming UCT messages without a UCT descriptor */
+        status = ucs_mpool_set_init(&worker->am_mps,
+                                    context->config.am_mpools.sizes,
+                                    context->config.am_mpools.count,
+                                    max_mp_entry_size, 0,
+                                    UCP_WORKER_HEADROOM_SIZE + worker->am.alignment,
+                                    0, UCS_SYS_CACHE_LINE_SIZE, 128, UINT_MAX,
+                                    &ucp_am_mpool_ops, "ucp_am_bufs");
+        if (status != UCS_OK) {
+            goto err_reg_mp_cleanup;
+        }
+    }
+
+    worker->max_uct_am_size = max_mp_entry_size;
```

Comment:
maybe use a flag on worker instead?

Reply:
maybe this value can be useful for something else in the future?

Reply:
seems it would already be saved in the mpool  set

Reply:
imo, it does not look good to get it from there, it is just internals of specific data struct.

Reply:
also, it's not really max_uct_am_size 
it's max possible **received** active message length
IMO better remote and add if really needed

Reply:
` possible received active message length`  - not sure if this make sense either